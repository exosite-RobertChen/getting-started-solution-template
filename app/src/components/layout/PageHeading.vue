<template>
  <div
    class="mt-4 pr-2 pl-6 page-heading"
    :class="{ 'mb-2': $vuetify.breakpoint.xs, 'mb-4': !$vuetify.breakpoint.xs }"
  >
    <h1 class="hidden-sm-and-up headline mt-4">{{ name }}</h1>
    <v-row align="center">
      <h1 class="hidden-xs-only mr-5">{{ name }}</h1>
      <slot />
    </v-row>
  </div>
</template>

<script>
export default {
  props: {
    name: {
      required: true,
      type: String
    }
  }
}
</script>

<style lang="scss">
.page-heading {
  h1 {
    font-weight: 300;
  }

  @media only screen and (min-width: 601px) {
    min-height: 60px;
    display: flex;
    align-items: flex-end;
  }
}
</style>
