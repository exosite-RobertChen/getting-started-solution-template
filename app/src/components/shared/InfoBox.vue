<template>
  <v-row
    :class="{ 'subtitle-1': $vuetify.breakpoint.xs }"
    class="headline fill-height"
    justify="center"
    align="center"
  >
    <v-card
      :class="{
        'px-6': $vuetify.breakpoint.xs,
        'px-10': !$vuetify.breakpoint.xs
      }"
      class="center py-10 text-center elevation-5 border-radius-1"
    >
      <div
        v-for="text in texts"
        :key="text"
        class="mb-2"
        :class="{ 'accent--text': error }"
      >
        {{ text }}
      </div>
      <div v-if="btnText">
        <span>Go ahead and</span>
        <v-btn
          class="ml-3"
          color="accent"
          :small="$vuetify.breakpoint.xs"
          data-qa="InfoBoxButton"
          @click="$emit('action')"
          >{{ btnText }}</v-btn
        >
      </div>
      <div class="mt-6">
        <a
          :href="links.help"
          class="subtitle-1 info-box__link"
          target="_blank"
          primary--text
          >Need help?</a
        >
      </div>
    </v-card>
  </v-row>
</template>

<script>
export default {
  name: 'InfoBox',
  props: {
    texts: {
      type: Array,
      required: true
    },
    btnText: {
      type: String,
      default: ''
    },
    links: {
      type: Object,
      required: true
    },
    error: {
      type: Boolean
    }
  }
}
</script>

<style lang="scss">
.info-box {
  &__link {
    text-decoration: underline;
  }
}
</style>
