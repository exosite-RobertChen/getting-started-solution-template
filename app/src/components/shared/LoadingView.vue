<template>
  <v-row justify="center" align="center">
    <v-col class="text-center">
      <v-progress-circular :size="50" color="primary" indeterminate />
      <div class="mt-6" :class="textColorClass">{{ text }}</div>
    </v-col>
  </v-row>
</template>

<script>
export default {
  props: {
    text: {
      type: String,
      required: false,
      default: ''
    },
    textColorClass: {
      type: String,
      required: false,
      default: ''
    }
  }
}
</script>
