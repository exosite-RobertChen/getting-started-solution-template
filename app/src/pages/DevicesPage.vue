<template>
  <v-container style="height: 100%">
    <PageHeading name="Devices">
      <v-tooltip v-model="showTooltip" top :disabled="!!selected.length">
        <template v-slot:activator="{ on }">
          <div v-on="on">
            <v-btn
              class="ml-4"
              outlined
              :disabled="!selected.length"
              @click="openAppConfigurationDialog"
            >
              <v-icon left>edit</v-icon>Configure
            </v-btn>
          </div>
        </template>
        <span>Select a device to configure</span>
      </v-tooltip>

      <v-spacer class="hidden-xs-only" />
      <v-col cols="5" md="3" class="mr-4">
        <v-text-field
          v-model="search"
          append-icon="search"
          label="Search"
          single-line
          class="standard-search"
        />
      </v-col>
      <v-col cols="5" md="3" lg="2" class="ml-auto">
        <v-select
          v-model="selectedAppId"
          :items="Object.values(apps)"
          :clearable="true"
          label="Filter by app"
          item-text="name"
          item-value="id"
          outlined
          class="standard-select"
        />
      </v-col>
    </PageHeading>

    <!-- main content -->
    <!-- mobile-breakpoint is minimal as it is not used -->
    <v-data-table
      v-model="selected"
      :headers="headers"
      :items="devices"
      :search="search"
      item-key="identity"
      :page.sync="pageSelect"
      :server-items-length="serverItems"
      :loading="devicesLoading"
      loading-text="Loading Devices"
      :footer-props="{
        'items-per-page-options': [5, 10, 20, 50, 100]
      }"
      :items-per-page.sync="perPageItems"
      must-sort
      show-select
      :mobile-breakpoint="200"
      class="elevation-4 border-radius-1 overflow-hidden"
      data-qa="DevicesTable"
      @click:row="openDeviceDetailsDialog"
    >
      <template v-slot:item.lastip="{ item }">
        {{ item.lastip || '-' }}
      </template>
      <template v-slot:item.apps="{ item }">
        {{ (item.apps || []).length }}
      </template>
      <template v-slot:item.update="{ item }">
        {{
          item.lastseen === 0
            ? `Hasn't reported yet`
            : new Date(item.lastseen / 1000).toLocaleString()
        }}
      </template>
      <template v-slot:item.online="{ item }">
        <v-chip v-if="item.online" class="chip-online">online</v-chip>
        <span v-else class="pl-1" color="grey">offline</span>
      </template>
      <template v-slot:no-results>
        <v-alert color="error" icon="warning"
          >Your search for "{{ search }}" found no results.</v-alert
        >
      </template>
    </v-data-table>

    <!-- lazy loading (not implemented yet) -->
    <!-- <v-row v-if="canLoadMore" class="text-center mt-4" justify="center">
      <v-btn outlined @click="loadSomeMore">Load more devices</v-btn>
    </v-row> -->

    <!-- special cases -->
    <v-row
      v-if="showInfoBox"
      justify="center"
      align="center"
      style="height: 60%"
    >
      <InfoBox
        v-if="showInfoBox"
        :texts="errors || ['There are no devices for this app']"
        :btn-text="!errors ? 'Claim Device' : null"
        :links="branding.links"
        :error="!!errors"
        @action="$emit('openClaimDevice')"
      />
    </v-row>

    <!-- dialogs -->
    <DeviceDetailsDialog
      v-model="deviceDetailsDialogOpen"
      :device="selectedDevice"
      :branding="branding"
      :apps="apps"
      @openResetDeviceDialog="openResetDeviceDialog"
      @app-configure-done="appConfigureDone"
      @app-to-add="appToAdd"
    />
    <ResetDeviceDialog
      v-model="resetDeviceDialogOpen"
      :device="selectedDevice"
      :branding="branding"
      @openSnackbar="$emit('openSnackbar')"
      @action-done="fetchDevices"
    />
    <AppConfigurationDialog
      v-model="appConfigurationDialogOpen"
      :apps="apps"
      :devices="selected"
      :branding="branding"
      @openSnackbar="$emit('openSnackbar')"
      @action-done="appConfigureDone"
    />
  </v-container>
</template>

<script>
import InfoBox from '@/components/shared/InfoBox'
import DeviceDetailsDialog from '@/components/devices/DeviceDetailsDialog'
import ResetDeviceDialog from '@/components/devices/ResetDeviceDialog'
import PageHeading from '@/components/layout/PageHeading'
import deviceapi from '@/api/devices'
import AppConfigurationDialog from '@/components/devices/AppConfigurationDialog'

export default {
  components: {
    AppConfigurationDialog,
    InfoBox,
    DeviceDetailsDialog,
    PageHeading,
    ResetDeviceDialog
  },
  props: {
    devicesClaimed: {
      type: Number,
      required: true
    },
    apps: {
      type: Object,
      required: true
    },
    globalError: {
      type: String,
      required: true
    },
    branding: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      appConfigurationDialogOpen: false,
      devicesLoading: true,
      search: '',
      devices: [],
      selected: [],
      localError: null,
      selectedAppId: '',
      selectedDevice: { state: {} },
      deviceDetailsDialogOpen: false,
      resetDeviceDialogOpen: false,
      showTooltip: false,
      serverItems: 0,
      pageSelect: 1,
      perPageItems: 10
    }
  },
  computed: {
    errors() {
      const result = []
      if (this.globalError) {
        result.push(this.globalError)
      }
      if (this.localError) {
        result.push(this.localError)
      }

      return result.length > 0 ? result : null
    },
    showInfoBox() {
      return this.errors
    },
    headers() {
      if (this.$vuetify.breakpoint.xs) {
        return [
          { text: 'Device ID', value: 'identity', sortable: true },
          { text: 'Status', value: 'online', sortable: true }
        ]
      } else {
        return [
          { text: 'Device ID', value: 'identity', sortable: true },
          { text: 'Status', value: 'online', sortable: true },
          { text: 'Last IP', value: 'lastip', sortable: true },
          { text: 'Last Seen', value: 'update', sortable: true },
          { text: 'Apps', value: 'apps', sortable: true }
        ]
      }
    }
  },
  watch: {
    perPageItems() {
      //wont modify count, just fetch again
      this.fetchDevices()
    },
    pageSelect() {
      //wont modify count, just fetch again
      this.fetchDevices()
    },
    devicesClaimed() {
      this.fetchDevices()
    },
    selectedAppId() {
      //can modify count devices
      this.selected = []
      this.countAllDevices()
    },
    search() {
      //can modify count devices
      //introduce a debouncer to avoid fetch too many times
      this.selected = []
      this.delay(this.countAllDevices(), 500)
    }
  },
  created() {
    this.handleDeviceperPage()
  },
  mounted() {
    this.assignFilter()
    this.countAllDevices()
  },
  methods: {
    appConfigureDone() {
      this.fetchDevices()
      this.selected = []
    },
    openAppConfigurationDialog() {
      this.appConfigurationDialogOpen = true
    },
    async fetchDevices() {
      this.devicesLoading = true
      //get limit and offset, offset is taken from options depending page selected
      const { error, data } = await deviceapi.fetchDevices(
        this.selectedAppId,
        this.search,
        this.perPageItems,
        (this.pageSelect - 1) * this.perPageItems
      )

      if (error) {
        this.$emit('openSnackbar', 'error', error)
        return (this.localError = error.toString())
      } else {
        this.devices = data.devices
      }

      this.devicesLoading = false
    },
    openDeviceDetailsDialog(device) {
      this.selectedDevice = device
      this.deviceDetailsDialogOpen = true
    },
    openResetDeviceDialog() {
      this.resetDeviceDialogOpen = true
    },
    appToAdd() {
      this.selected = [this.selectedDevice]
      this.appConfigurationDialogOpen = true
    },
    assignFilter() {
      if (this.$route && this.$route.params.select) {
        this.selectedAppId = this.$route.params.select
      }
    },
    async countAllDevices() {
      //Detect number of devices, depending seleccted App or Search
      //Result used in pagination footer
      //will call then fetchDevices
      const { error, counter } = await deviceapi.countDevices(
        this.selectedAppId,
        this.search
      )
      if (error) {
        this.$emit('openSnackbar', 'error', error)
        return (this.localError = error.toString())
      } else {
        this.serverItems = counter
        this.fetchDevices()
      }
    },
    delay(callback, ms) {
      var timer = 0
      return function() {
        var context = this,
          args = arguments
        clearTimeout(timer)
        timer = setTimeout(function() {
          callback.apply(context, args)
        }, ms || 0)
      }
    },
    handleDeviceperPage() {
      if (this.$vuetify.breakpoint.height < 720) {
        this.perPageItems = 5
      } else {
        this.perPageItems = 10
      }
    }
  }
}
</script>

<style lang="scss">
.api-result {
  background-color: #777;
  color: white;
}

.border-right {
  border-right: 1px solid rgba(0, 0, 0, 0.12);
}
.v-window {
  height: 100% !important;
}

.border-radius-1 {
  border-radius: 10px !important;
}
.border-radius-0 {
  border-radius: 5px !important;
}
.overflow-hidden {
  overflow: hidden;
}

tr {
  cursor: pointer;
}

.grey-bg {
  background-color: rgb(233, 233, 233);
}
.jsontree_bg {
  background: #fff;
}
</style>
