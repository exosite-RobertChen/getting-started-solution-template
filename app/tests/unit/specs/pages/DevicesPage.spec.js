import merge from 'lodash.merge'

import { mountComponent, shallowComponent } from '@/../tests/unit/common'
import branding from '@/../vendor/branding.json'
import DevicesPage from '@/pages/DevicesPage'
import LoadingView from '@/components/shared/LoadingView'
import InfoBox from '@/components/shared/InfoBox'
import PageHeading from '@/components/layout/PageHeading'

const defaultProps = {
  sync: false,
  propsData: {
    branding,
    devicesClaimed: 0,
    globalError: '',
    apps: {}
  }
}

const shallowFactory = overrides =>
  shallowComponent(DevicesPage, merge(defaultProps, overrides))

const mountFactory = overrides =>
  mountComponent(DevicesPage, merge(defaultProps, overrides))

describe('DevicesPage.vue', () => {
  describe('when devices are loading', () => {
    const wrapper = shallowFactory()

    test('can be mounted', () => {
      expect(wrapper.contains(DevicesPage)).toBe(true)
    })

    test('renders only heading and spinner initially', async () => {
      expect(wrapper.contains(PageHeading)).toBe(true)
      expect(wrapper.contains(InfoBox)).toBe(false)
      await wrapper.vm.$forceUpdate()
      expect(wrapper.contains('[data-qa=DevicesTable]')).toBe(true)
    })

    test('there are no devices shows blank devicetable', async ()=> {
      wrapper.vm.devicesLoading = false
      expect(wrapper.contains('[data-qa=DevicesTable]')).toBe(true)
      expect(wrapper.contains(InfoBox)).toBe(false)
    })

    test('there are error shows infoxbox', async ()=> {
      wrapper.vm.globalError = 'No device due to error'
      await wrapper.vm.$nextTick()
      await wrapper.vm.$forceUpdate()
      expect(wrapper.contains(InfoBox)).toBe(true)
    })
  })

  describe('when there are devices loaded', () => {
    const wrapper = mountFactory()
    wrapper.setMethods({ fetchDevices: jest.fn(),delay: jest.fn(),countAllDevices: jest.fn() ,handleDeviceperPage: jest.fn()})
    wrapper.setProps({
      headers: [
        { text: 'Device ID', value: 'identity', sortable: true },
          { text: 'Status', value: 'online', sortable: true },
          { text: 'Last IP', value: 'lastip', sortable: true },
          { text: 'Last Seen', value: 'update', sortable: true },
          { text: 'Apps', value: 'apps', sortable: true }
      ],
      apps: {
        adsfhgdas: {
          name: 'app1',
          tr_st: 200,
          default: true,
          type: 1
        },
        myAwesomeId: {
          name: 'app2',
          tr_st: 401,
          default: false,
          type: 1,
          description:
            'dfasilhgads flahksfd aslhdkflashkd flakhsdflkah sdlfhas dhaoighsh-ph3-qhrreoghaobh asd'
        }
      }
    })

    wrapper.setData({
      serverItems: 2,
      pageSelect: 1,
      perPageItems: 10,
      devicesLoading: false,
      devices: [
        {
          identity: 'device1',
          online: true,
          lastIp: '255.255.255.255',
          lastSeen: 12343344,
          state: {},
          group: {
            id: 'groupOneId'
          }
        },
        {
          identity: 'device2',
          online: true,
          lastIp: '245.245.255.255',
          lastSeen: 30983433,
          state: {},
          group: {
            id: 'groupOneId'
          }
        }
      ]
    })

    test('renders a device table and a row for each device once data is loaded',  () => {
      //await wrapper.vm.$forceUpdate()
      expect(wrapper.contains('table')).toBe(true)
      //the first tr is for heading, the 2 others for 2 devices on list
      expect(wrapper.findAll('tr').length).toBe(3)
    })

    test('matches snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    test('opens device details dialog if a device row is clicked', async () => {
      wrapper.setMethods({ openDeviceDetailsDialog: jest.fn() })
      expect(wrapper.vm.openDeviceDetailsDialog).not.toHaveBeenCalled()
      wrapper
        .findAll('tr')
        .at(1)
        .trigger('click')
      expect(wrapper.vm.openDeviceDetailsDialog).toHaveBeenCalled()
    })

    test('when selected appId changes, countAllDevices is called again', async () => {
      wrapper.setMethods({ countAllDevices: jest.fn() })
      expect(wrapper.vm.countAllDevices).toHaveBeenCalledTimes(0)

      wrapper.vm.$options.watch.selectedAppId.call(wrapper.vm)
      wrapper.setData({
        selectedAppId: 'foo bar'
      })
      expect(wrapper.vm.countAllDevices).toHaveBeenCalledTimes(1)
    })
  })
})
