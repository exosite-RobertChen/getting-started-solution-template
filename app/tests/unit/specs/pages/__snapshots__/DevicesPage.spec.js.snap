// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevicesPage.vue when there are devices loaded matches snapshot 1`] = `
<div
  class="container"
  style="height: 100%;"
>
  <div
    class="mt-4 pr-2 pl-6 page-heading mb-4"
  >
    <h1
      class="hidden-sm-and-up headline mt-4"
    >
      Devices
    </h1>
     
    <div
      class="row align-center"
    >
      <h1
        class="hidden-xs-only mr-5"
      >
        Devices
      </h1>
       
      <span
        class="v-tooltip v-tooltip--top"
      >
        <div
          class="v-tooltip__content"
          style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"
        />
      </span>
      <div>
        <button
          class="ml-4 v-btn v-btn--depressed v-btn--disabled v-btn--flat v-btn--outlined theme--light v-size--default"
          disabled="disabled"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate v-icon--left material-icons theme--light"
            >
              edit
            </i>
            Configure
          
          </span>
        </button>
      </div>
       
      <div
        class="spacer hidden-xs-only"
      />
       
      <div
        class="mr-4 col-md-3 col-5"
      >
        <div
          class="v-input standard-search theme--light v-text-field v-text-field--single-line"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-19"
                  style="left: 0px; position: absolute;"
                >
                  Search
                </label>
                <input
                  id="input-19"
                  type="text"
                />
              </div>
              <div
                class="v-input__append-inner"
              >
                <div
                  class="v-input__icon v-input__icon--append"
                >
                  <i
                    aria-hidden="true"
                    class="v-icon notranslate material-icons theme--light"
                  >
                    search
                  </i>
                </div>
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <div
                  class="v-messages__wrapper"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="ml-auto col-md-3 col-lg-2 col-5"
      >
        <div
          class="v-input standard-select theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select"
        >
          <div
            class="v-input__control"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="list-23"
              class="v-input__slot"
              role="button"
            >
              <fieldset
                aria-hidden="true"
              >
                <legend
                  style="width: 0px;"
                >
                  <span>
                    â€‹
                  </span>
                </legend>
              </fieldset>
              <div
                class="v-select__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-23"
                  style="left: 0px; position: absolute;"
                >
                  Filter by app
                </label>
                <div
                  class="v-select__selections"
                >
                  <input
                    aria-readonly="true"
                    id="input-23"
                    readonly="readonly"
                    type="text"
                  />
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--"
                  >
                    <i
                      class="v-icon notranslate v-icon--link material-icons theme--light"
                      role="button"
                    />
                  </div>
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--append"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-menu-down theme--light"
                    />
                  </div>
                </div>
              </div>
              <div
                class="v-menu"
              >
                <div
                  class="v-menu__content theme--light "
                  style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <div
                  class="v-messages__wrapper"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="v-data-table elevation-4 border-radius-1 overflow-hidden theme--light"
    data-qa="DevicesTable"
  >
    <div
      class="v-data-table__wrapper"
    >
      <table>
        <colgroup>
          <col
            class=""
          />
          <col
            class=""
          />
          <col
            class=""
          />
          <col
            class=""
          />
          <col
            class=""
          />
          <col
            class=""
          />
        </colgroup>
        <thead
          class="v-data-table-header"
        >
          <tr>
            <th
              aria-label=""
              aria-sort="none"
              class="text-start"
              role="columnheader"
              scope="col"
              style="width: 1px; min-width: 1px;"
            >
              <div
                class="v-data-table__checkbox v-simple-checkbox"
              >
                <div
                  class="v-input--selection-controls__ripple"
                />
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"
                />
              </div>
            </th>
            <th
              aria-label="Device ID: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="text-start sortable"
              role="columnheader"
              scope="col"
            >
              <span>
                Device ID
              </span>
              <i
                aria-hidden="true"
                class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                style="font-size: 18px;"
              />
            </th>
            <th
              aria-label="Status: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="text-start sortable"
              role="columnheader"
              scope="col"
            >
              <span>
                Status
              </span>
              <i
                aria-hidden="true"
                class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                style="font-size: 18px;"
              />
            </th>
            <th
              aria-label="Last IP: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="text-start sortable"
              role="columnheader"
              scope="col"
            >
              <span>
                Last IP
              </span>
              <i
                aria-hidden="true"
                class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                style="font-size: 18px;"
              />
            </th>
            <th
              aria-label="Last Seen: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="text-start sortable"
              role="columnheader"
              scope="col"
            >
              <span>
                Last Seen
              </span>
              <i
                aria-hidden="true"
                class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                style="font-size: 18px;"
              />
            </th>
            <th
              aria-label="Apps: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="text-start sortable"
              role="columnheader"
              scope="col"
            >
              <span>
                Apps
              </span>
              <i
                aria-hidden="true"
                class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                style="font-size: 18px;"
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
          >
            <td
              class="text-start"
            >
              <div
                class="v-data-table__checkbox v-simple-checkbox"
              >
                <div
                  class="v-input--selection-controls__ripple"
                />
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"
                />
              </div>
            </td>
            <td
              class="text-start"
            >
              device1
            </td>
            <td
              class="text-start"
            >
              <span
                class="chip-online v-chip v-chip--no-color theme--light v-size--default"
              >
                <span
                  class="v-chip__content"
                >
                  online
                </span>
              </span>
            </td>
            <td
              class="text-start"
            >
              
      -
    
            </td>
            <td
              class="text-start"
            >
              
      Invalid Date
    
            </td>
            <td
              class="text-start"
            >
              
      0
    
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="text-start"
            >
              <div
                class="v-data-table__checkbox v-simple-checkbox"
              >
                <div
                  class="v-input--selection-controls__ripple"
                />
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"
                />
              </div>
            </td>
            <td
              class="text-start"
            >
              device2
            </td>
            <td
              class="text-start"
            >
              <span
                class="chip-online v-chip v-chip--no-color theme--light v-size--default"
              >
                <span
                  class="v-chip__content"
                >
                  online
                </span>
              </span>
            </td>
            <td
              class="text-start"
            >
              
      -
    
            </td>
            <td
              class="text-start"
            >
              
      Invalid Date
    
            </td>
            <td
              class="text-start"
            >
              
      0
    
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="v-data-footer"
    >
      <div
        class="v-data-footer__select"
      >
        Rows per page:
        <div
          class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select"
        >
          <div
            class="v-input__control"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="list-2387"
              class="v-input__slot"
              role="button"
            >
              <div
                class="v-select__slot"
              >
                <div
                  class="v-select__selections"
                >
                  <div
                    class="v-select__selection v-select__selection--comma"
                  >
                    10
                  </div>
                  <input
                    aria-label="$vuetify.dataTable.itemsPerPageText"
                    aria-readonly="true"
                    id="input-2387"
                    readonly="readonly"
                    type="text"
                  />
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--append"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-menu-down theme--light"
                    />
                  </div>
                </div>
              </div>
              <div
                class="v-menu"
              >
                <div
                  class="v-menu__content theme--light "
                  style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="v-data-footer__pagination"
      >
        1-2 of 2
      </div>
      <div
        class="v-data-footer__icons-before"
      >
        <button
          aria-label="Previous page"
          class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
          disabled="disabled"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-chevron-left theme--light"
            />
          </span>
        </button>
      </div>
      <div
        class="v-data-footer__icons-after"
      >
        <button
          aria-label="Next page"
          class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
          disabled="disabled"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-chevron-right theme--light"
            />
          </span>
        </button>
      </div>
    </div>
  </div>
   
  <!---->
   
  <div
    class="v-dialog__container"
    role="dialog"
  >
    <div
      class="v-dialog__content"
      role="document"
      style="z-index: 0;"
    >
      <div
        class="v-dialog"
        style="max-width: 600px; display: none;"
      />
    </div>
  </div>
   
  <div
    class="v-dialog__container"
    role="dialog"
  >
    <div
      class="v-dialog__content"
      role="document"
      style="z-index: 0;"
    >
      <div
        class="v-dialog"
        style="max-width: 600px; display: none;"
      />
    </div>
  </div>
   
  <div
    class="v-dialog__container"
    role="dialog"
  >
    <div
      class="v-dialog__content"
      role="document"
      style="z-index: 0;"
    >
      <div
        class="v-dialog"
        style="max-width: 700px; display: none;"
      />
    </div>
  </div>
</div>
`;
